monty-31/src/aarch64_neon/packing.rs

```diff
@@ -586,38 +591,261 @@ impl_packed_field_pow_2!(
     WIDTH
 );
 
+/// Compute the lane-wise widening product of two 128-bit vectors of `u32`.
+///
+/// Input vectors are
+///
+/// \begin{equation}
+///     v_a = [a_0, a_1, a_2, a_3], \qquad
+///     v_b = [b_0, b_1, b_2, b_3].
+/// \end{equation}
+///
+/// The result is two 64-bit vectors:
+///
+/// \begin{equation}
+///     lo = [a_0 ⋅ b_0, a_1 ⋅ b_1], \qquad
+///     hi = [a_2 ⋅ b_2, a_3 ⋅ b_3],
+/// \end{equation}
+///
+/// where each product is the exact 32×32 → 64-bit integer product (no reduction).
+#[inline(always)]
+unsafe fn vmull_wide_u32(a: uint32x4_t, b: uint32x4_t) -> (uint64x2_t, uint64x2_t) {
+    unsafe {
+        // We first extract those low halves from a and b:
+        //
+        // [a_0, a_1] as u32x2
+        let a_lo = aarch64::vget_low_u32(a);
+        // [b_0, b_1] as u32x2
+        let b_lo = aarch64::vget_low_u32(b);
+
+        // lo = [a_0 ⋅ b_0, a_1 ⋅ b_1] as u64 lanes (exact 32×32 → 64 multiply, unsigned).
+        let lo = aarch64::vmull_u32(a_lo, b_lo);
+
+        // hi = [a_2 ⋅ b_2, a_3 ⋅ b_3] as u64 lanes.
+        let hi = aarch64::vmull_high_u32(a, b);
+
+        (lo, hi)
+    }
+}
+
+/// A general fast dot product implementation using NEON.
+#[inline(always)]
+fn general_dot_product<P, const N: usize>(
+    lhs: &[MontyField31<P>],
```

Comment:
Eventually this should accept either a `MontyField31<P>` or a `PackedMontyField31Neon<P>` but we can do this in a future PR. 
```suggestion
    lhs: &[MontyField31<P>], // TODO: Support MontyField31<P> or PackedMontyField31Neon<P>.
```